# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/IdleRequest.cpp
# Commit: be6c223e63f0
# Full Hash: be6c223e63f0a0c8ef2428df42321d59e4c0f7c6
# Author: Sean Feng <sefeng@mozilla.com>
# Date: 2025-02-25 09:30:56
# Regressor Bug: 1920115
# File Overlap Count: 2
# Description:
#   Bug 1920115 - Implement scheduler.yield r=jjaschke,dom-core,webidl,smaug
#   
#   Spec: https://wicg.github.io/scheduling-apis/#dom-scheduler-yield
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D234232
# ==============================================================================

diff -r 121bf804d2a6 -r be6c223e63f0 dom/base/IdleRequest.cpp
--- a/dom/base/IdleRequest.cpp	Mon Feb 24 19:31:26 2025 +0000
+++ b/dom/base/IdleRequest.cpp	Mon Feb 24 19:31:27 2025 +0000
@@ -13,6 +13,7 @@
 #include "mozilla/dom/WindowBinding.h"
 #include "nsComponentManagerUtils.h"
 #include "nsPIDOMWindow.h"
+#include "mozilla/dom/WebTaskScheduler.h"
 
 namespace mozilla::dom {
 
@@ -54,7 +55,20 @@
       new IdleDeadline(aWindow, aDidTimeout, aDeadline);
   RefPtr<IdleRequestCallback> callback(std::move(mCallback));
   MOZ_ASSERT(!mCallback);
+
+  RefPtr<nsGlobalWindowInner> innerWindow = nsGlobalWindowInner::Cast(aWindow);
+  // https://wicg.github.io/scheduling-apis/#sec-patches-invoke-idle-callbacks
+  // Let state be a new scheduling state.
+  RefPtr<WebTaskSchedulingState> newState = new WebTaskSchedulingState();
+  // Set state’s priority source to the result of creating a fixed priority
+  // unabortable task signal given "background" and realm.
+  newState->SetPrioritySource(
+      new TaskSignal(aWindow->AsGlobal(), TaskPriority::Background));
+  // Set event loop’s current scheduling state to state.
+  innerWindow->SetWebTaskSchedulingState(newState);
   callback->Call(*deadline, "requestIdleCallback handler");
+  // Set event loop’s current scheduling state to null.
+  innerWindow->SetWebTaskSchedulingState(nullptr);
 }
 
 }  // namespace mozilla::dom